src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs

```diff
@@ -0,0 +1,196 @@
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// Helper type for building dynamically-sized arrays while minimizing allocations and copying.
+    /// </summary>
+    internal struct LargeArrayBuilder<T>
+    {
+        private const int StartingCapacity = 4;
+        private const int ResizeLimit = 32;
+
+        private T[] _first;                 // The first buffer we store items in. Resized until ResizeLimit.
+        private ArrayBuilder<T[]> _buffers; // After ResizeLimit * 2, we store previous buffers we've filled out here.
+        private T[] _current;               // Current buffer we're reading into. Is _first <= ResizeLimit.
+        private int _index;                 // Index into the current buffer.
+        private int _count;                 // Count of all of the items in this builder.
+
+        /// <summary>
+        /// Constructs a new builder.
+        /// </summary>
+        /// <param name="initialize">Pass <c>true</c>.</param>
+        public LargeArrayBuilder(bool initialize) : this()
+        {
+            // This is a workaround for C# not having parameterless struct constructors yet.
+            // Once it gets them, replace this with a parameterless constructor.
+            Debug.Assert(initialize);
+
+            _first = _current = Array.Empty<T>();
+        }
+
+        /// <summary>
+        /// Gets the number of items added to the builder.
+        /// </summary>
+        public int Count => _count;
+
+        /// <summary>
+        /// Adds an item to this builder.
+        /// </summary>
+        /// <param name="item">The item to add.</param>
+        /// <remarks>
+        /// Use <see cref="Add"/> if adding to the builder is a bottleneck for your use case.
+        /// Otherwise, use <see cref="SlowAdd"/>.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Add(T item)
+        {
+            if (_index == _current.Length)
+            {
+                AllocateBuffer();
+            }
+
+            _current[_index++] = item;
+            _count++;
+        }
+
+        /// <summary>
+        /// Adds a range of items to this builder.
+        /// </summary>
+        /// <param name="items">The sequence to add.</param>
+        public void AddRange(IEnumerable<T> items)
+        {
+            Debug.Assert(items != null);
+
+            using (IEnumerator<T> enumerator = items.GetEnumerator())
+            {
+                T[] destination = _current;
+                int index = _index;
+
+                // Continuously read in items from the enumerator, updating _count
+                // and _index when we run out of space.
+
+                while (enumerator.MoveNext())
+                {
+                    if (index == destination.Length)
+                    {
+                        // No more space in this buffer. Resize.
+                        _count += index - _index;
```

Comment:
Ah, very nice catch. It looks like we will multiply `_count` by 2 again, and 0x80000000 * 2 is 0, so then we allocate a 0-length array and set it to `_first` / `_current` instead of raising an exception. Since it doesn't add any additional cost, I can change the check to be `if ((uint)_count < (uint)ResizeLimit)` which should get us down the path which throws for overflow.